packages/shared_preferences/shared_preferences_ios/ios/Classes/messages.g.h

```diff
@@ -0,0 +1,33 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+// Autogenerated from Pigeon (v1.0.16), do not edit directly.
+// See also: https://pub.dev/packages/pigeon
+#import <Foundation/Foundation.h>
+@protocol FlutterBinaryMessenger;
+@protocol FlutterMessageCodec;
+@class FlutterError;
+@class FlutterStandardTypedData;
+
+NS_ASSUME_NONNULL_BEGIN
+
+/// The codec used by UserDefaultsApi.
+NSObject<FlutterMessageCodec> *UserDefaultsApiGetCodec(void);
+
+@protocol UserDefaultsApi
+- (void)removeKey:(NSString *)key error:(FlutterError *_Nullable *_Nonnull)error;
+- (void)setBoolKey:(NSString *)key
+             value:(NSNumber *)value
+             error:(FlutterError *_Nullable *_Nonnull)error;
+- (void)setDoubleKey:(NSString *)key
+               value:(NSNumber *)value
+               error:(FlutterError *_Nullable *_Nonnull)error;
+- (void)setValueKey:(NSString *)key value:(id)value error:(FlutterError *_Nullable *_Nonnull)error;
+- (nullable NSDictionary<NSString *, id> *)getAllWithError:(FlutterError *_Nullable *_Nonnull)error;
```

Comment:
Why is Pigeon annotating this as nullable? It's non-nullable in the Dart specification; that seems like a significant Pigeon bug. (Or is this stale?)

Reply:
It's not stale.  It was probably done because of error handling.  If there is an error you don't want people to artificially return some value.

Reply:
Yea, reading through the code, that's what it's for.  If you return nil and set an error, it will throw a `PlatformException` on the dart side.  That's what we want.  If you return nil and don't set an error, you'll get an exception when it is cast to a non-null type before returning.  I think that's what we want as well.

Reply:
That means that it's currently impossible to distinguish in the context of the native code between a nullable and a non-nullable return value, which is definitely not great. We can't do anything compiler-enforceable I guess, which is unfortunate, but it seems like we should at least have pigeon put a comment on methods like this that says:
```
/// This may only return nil if 'error' is set.
```

so that there's clear information within the context of the generated native code what the required behavior is.

> If you return nil and don't set an error, you'll get an exception when it is cast to a non-null type before returning. I think that's what we want as well.

It looks like that's only on the Dart side. Shouldn't we be validating in ObjC code that either `error` or the return value is non-nil, in the case of non-nullable return values? Otherwise mistakes will look like they are on the Dart side rather than in the native implementation.

Reply:
Yea, those are improvements we can do when we implement nullable return types.  I'll add a note to that issue: https://github.com/flutter/flutter/issues/98452#issuecomment-1040718089

> Shouldn't we be validating in ObjC code that either error or the return value is non-nil, in the case of non-nullable return values? Otherwise mistakes will look like they are on the Dart side rather than in the native implementation.

It doesn't necessarily have to happen on the objc side.  The end result will look the same to a user, an exception on the Dart side.  I think generating a `PlatformException` about a null value instead of a null value exception will be more clear though.

> it seems like we should at least have pigeon put a comment on methods like this that says:

Yea, it's a good idea.  I don't know how helpful it will be in practice because it won't show up where people are actually implementing the logic.

Reply:
> The end result will look the same to a user, an exception on the Dart side.

Right, I wasn't thinking of the end user, but the plugin developer(s). Especially if multiple platforms share method channels (which is still overwhelmingly the common case in the ecosystem). If the exception is in the Dart code, it points someone investigating in the wrong direction; probably not a huge effect, but anything that makes it harder for people to root-cause bugs isn't ideal.